city_scrapers/spiders/pitt_housing.py

```diff
@@ -0,0 +1,145 @@
+                'all_day': False,
+                'location': {
+                    'address': 'Civic Building, 200 Ross St., Pittsburgh, PA, 15219',
+                    'name': '',
+                    'neighborhood': ''
+                },
+                'documents': self._parse_documents(item),
+                'sources': self._parse_sources(item),
+            }
+
+            data['status'] = self._generate_status(data, text='')
+            data['id'] = self._generate_id(data)
+
+            yield data
+
+    def _create_meetings(self, response):
+        """
+        Returns a list containing lists corresponding to each meeting.
+        """
+        meetings = []
+        parser = MyHTMLParser()
+        meetings_xpath = '//div[@id="the-copy"]/div[2]/p[contains(strong, "2018 HACP Board Agendas")]/following-sibling::p'
+
+        for item in response.xpath(meetings_xpath).extract():  
+            parser.feed(item)
+            meeting = [parser.date, [parser.doc_url]]
+
+            # If meetings is empty, append meeting
+            if not meetings:
+                meetings.append(meeting)
+
+            # If meetings is not empty, get the last meeting date and compare 
+            # it to parser.date. Append parser.doc_url to last meeting if 
+            # parser.date == the last meeting date. Otherwise, append the whole
+            # meeting to meetings.
+            else:
+                last_meeting = meetings[-1]
+                last_meeting_date = last_meeting[0]                
+                if parser.date == last_meeting_date:
+                    if parser.doc_url:
+                        last_meeting[1].append(parser.doc_url)  
+                else: 
+                    meetings.append(meeting)
+
+        return meetings
+
+    def _parse_start(self, item):
+        """
+        Parse start date and time.
+        """
+
+        date_string = item[0]
+        try:
+            dt = dtparse(date_string)
+            return {'date': dt.date(), 'time': None, 'note': None}
+        except ValueError:
+            return {'date': None, 'time': None, 'note': None}
+
+    def _parse_documents(self, item):
+        """
+        Parse or generate documents.
+        """
+
+        # Create documents, which will be a list of dictionaries corresponding 
+        # to each agenda or minutes document found.
+        documents = []
+        for doc in item[1]: 
+            if re.search(r'Agenda',doc):
+                agenda = {'url': doc, 'note': 'agenda'}
+                documents.append(agenda)
+            elif re.search(r'Minutes', doc):
+                minutes = {'url': doc, 'note': 'minutes'}
+                documents.append(minutes)
+
+        return documents
+
+    def _parse_sources(self, item):
+        """
+        Parse or generate sources.
+        """
+        return [{'url': self.start_urls[0], 'note': ''}]
+
+class MyHTMLParser(HTMLParser):
```

Comment:
Could you explain the purpose of this a bit? I'm not sure this needs to be split into a parse subclass, and it makes things a bit harder to maintain when there's a lot of variations across scrapers. If there is a good use case it might be worth seeing if this can be generalized for multiple scrapers (that doesn't need to be in this PR though)

Reply:
Sure thing.  Most meeting data were links and were formatted like this whenever scrapy grabbed their markup:

<p><a href="http://static1.firemandev.info.s3.amazonaws.com/galleries/general/2018%20Board%20Information/Board_Minutes_-_January_2018.pdf">Jan. 25, 2018 -- Minutes</a></p>

But those that were not links were formatted differently:

<p>July 31, 2018 -- Minutes</p>
 <p><strong>August 2018 -- Board does not meeting in August.\xa0</strong></p>

There was still quite a bit of HTML parsing to be done, so I used HTMLParser (which seemed like it can only be used through subclassing) to grab the links and dates, taking into account the differences in markup.  I was having trouble with keeping the events in chronological order, so this was the solution I came up with to grab everything we need in one pass.

Reply:
Makes sense! It doesn't add a huge amount of complexity, so if it works in this case I think that's fine

Reply:
Great, thanks for reviewing this!